src/MongoDB.Driver.Core/Core/Compression/ZlibCompressor.cs

```diff
@@ -0,0 +1,120 @@
+* limitations under the License.
+*/
+
+using System;
+using System.IO;
+using SharpCompress.Compressors;
+using SharpCompress.Compressors.Deflate;
+
+namespace MongoDB.Driver.Core.Compression
+{
+    /// <summary>
+    /// Compressor according to the zlib algorithm.
+    /// </summary> 
+    internal sealed class ZlibCompressor : ICompressor
+    {
+        private readonly CompressionLevel _compressionLevel;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ZlibCompressor" /> class.
+        /// </summary>
+        /// <param name="compressionLevel">The compression level.</param>
+        public ZlibCompressor(int? compressionLevel = 0)
+        {
+            _compressionLevel = GetCompressionLevel(compressionLevel);
+        }
+
+        /// <inheritdoc />
+        public CompressorType Type => CompressorType.Zlib;
+
+        /// <inheritdoc />
+        public ArraySegment<byte> Compress(ArraySegment<byte> bytes)
+        {
+            using (var memoryStream = new NonDisposingMemoryStream())
+            {
+                using (var zlibStream = new ZlibStream(memoryStream, CompressionMode.Compress, _compressionLevel))
+                {
+                    zlibStream.FlushMode = FlushType.Sync;
+                    zlibStream.Write(bytes.Array, bytes.Offset, bytes.Count);
+                }
+
+                memoryStream.AllowDispose = true;
+#if NETSTANDARD1_5
+                var buffer = memoryStream.ToArray();
+#else
+                var buffer = memoryStream.GetBuffer();
+#endif
+                return new ArraySegment<byte>(buffer, 0, (int)memoryStream.Length);
+            }
+        }
+
+        /// <inheritdoc />
+        public ArraySegment<byte> Decompress(ArraySegment<byte> bytes)
+        {
+            using (var inputStream = new MemoryStream(bytes.Array, bytes.Offset, bytes.Count))
+            using (var zlibStream = new ZlibStream(inputStream, CompressionMode.Decompress))
+            using (var outputStream = new MemoryStream())
+            {
+                zlibStream.CopyTo(outputStream);
+#if NETSTANDARD1_5
+                var buffer = outputStream.ToArray();
+#else
+                var buffer = outputStream.GetBuffer();
+#endif
+                return new ArraySegment<byte>(buffer, 0, (int)outputStream.Length);
+            }
+        }
+
+        private static CompressionLevel GetCompressionLevel(int? compressionLevel)
+        {
+            if (!compressionLevel.HasValue)
+            {
+                compressionLevel = -1;
+            }
+
+            switch (compressionLevel)
+            {
+                case -1:
+                    return CompressionLevel.Default;
+                case int _ when compressionLevel < -1 || compressionLevel > 9:
+                    throw new ArgumentOutOfRangeException(nameof(compressionLevel));
+                default:
+                    return (CompressionLevel)compressionLevel;
```

Comment:
How can a cast work? I thought we had to map from one set of numbers to a different set of numbers.